spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java

```diff
@@ -102,48 +139,54 @@ public void onMessage(Message message, Channel channel) {
 						getMessageBuilderFactory().withPayload(payload).copyHeaders(headers).build();
 				final org.springframework.messaging.Message<?> reply = sendAndReceiveMessage(request);
 				if (reply != null) {
-					// TODO: fallback to a reply address property of this gateway
 					Address replyTo;
 					String replyToProperty = message.getMessageProperties().getReplyTo();
-					// TODO: Use the Address.AMQ_RABBITMQ_REPLY_TO constant when 1.4.3 is the minimum
-					if (replyToProperty.startsWith("amq.rabbitmq.reply-to")) {
-						replyTo = new Address("", replyToProperty);
+					if (replyToProperty != null) {
+						replyTo = new Address(replyToProperty);
 					}
 					else {
-						replyTo = new Address(replyToProperty);
+						replyTo = AmqpInboundGateway.this.defaultReplyTo;
+					}
+
+					MessagePostProcessor messagePostProcessor = new MessagePostProcessor() {
+
+						@Override
+						public Message postProcessMessage(Message message) throws AmqpException {
+							MessageProperties messageProperties = message.getMessageProperties();
+							String contentEncoding = messageProperties.getContentEncoding();
+							long contentLength = messageProperties.getContentLength();
+							String contentType = messageProperties.getContentType();
+							headerMapper.fromHeadersToReply(reply.getHeaders(), messageProperties);
+							// clear the replyTo from the original message since we are using it now
+							messageProperties.setReplyTo(null);
+							// reset the content-* properties as determined by the MessageConverter
+							if (StringUtils.hasText(contentEncoding)) {
+								messageProperties.setContentEncoding(contentEncoding);
+							}
+							messageProperties.setContentLength(contentLength);
+							if (contentType != null) {
+								messageProperties.setContentType(contentType);
+							}
+							return message;
+						}
+
+					};
+
+					if (replyTo != null) {
+						amqpTemplate.convertAndSend(replyTo.getExchangeName(), replyTo.getRoutingKey(),
+								reply.getPayload(), messagePostProcessor);
+					}
+					else {
+						amqpTemplate.convertAndSend(reply.getPayload(),	messagePostProcessor);
```

Comment:
We should add some logic here - if not explicitly set, throw `IllegalStateException` - we know, for sure that the default template has no defaults for exchange/routing key.

We also need to clearly document the behavior - if you inject a template and don't set a default replyTo, the injected template's default exchange/routing key will be used to route replies.